src/Compression/CompressionCodecDeflate.cpp

```diff
@@ -0,0 +1,357 @@
+    // Performing a compression operation
+    job_ptr->op = qpl_op_compress;
+    job_ptr->next_in_ptr = reinterpret_cast<uint8_t *>(const_cast<char *>(source));
+    job_ptr->next_out_ptr = reinterpret_cast<uint8_t *>(dest);
+    job_ptr->available_in = source_size;
+    job_ptr->available_out = dest_size;
+    job_ptr->level = qpl_default_level;
+    job_ptr->flags = QPL_FLAG_FIRST | QPL_FLAG_DYNAMIC_HUFFMAN | QPL_FLAG_LAST | QPL_FLAG_OMIT_VERIFY;
+
+    // Compression
+    status = qpl_execute_job(job_ptr);
+    if (status != QPL_STS_OK)
+    {
+        throw Exception(
+            "SoftwareCodecDeflate::doCompressData -> qpl_execute_job fail:" + std::to_string(status), ErrorCodes::CANNOT_COMPRESS);
+    }
+
+    return job_ptr->total_out;
+}
+
+void SoftwareCodecDeflate::doDecompressData(const char * source, uint32_t source_size, char * dest, uint32_t uncompressed_size)
+{
+    qpl_status status;
+    qpl_job * job_ptr = getJobCodecPtr();
+
+    // Performing a decompression operation
+    job_ptr->op = qpl_op_decompress;
+    job_ptr->next_in_ptr = reinterpret_cast<uint8_t *>(const_cast<char *>(source));
+    job_ptr->next_out_ptr = reinterpret_cast<uint8_t *>(dest);
+    job_ptr->available_in = source_size;
+    job_ptr->available_out = uncompressed_size;
+    job_ptr->flags = QPL_FLAG_FIRST | QPL_FLAG_LAST;
+
+    // Decompression
+    status = qpl_execute_job(job_ptr);
+    if (status != QPL_STS_OK)
+    {
+        throw Exception(
+            "SoftwareCodecDeflate::doDecompressData -> qpl_execute_job fail:" + std::to_string(status), ErrorCodes::CANNOT_DECOMPRESS);
+    }
+}
+
+//CompressionCodecDeflate
+CompressionCodecDeflate::CompressionCodecDeflate()
+{
+    setCodecDescription("DEFLATE");
+    hwCodec = std::make_unique<HardwareCodecDeflate>();
+    swCodec = std::make_unique<SoftwareCodecDeflate>();
+}
+
+uint8_t CompressionCodecDeflate::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::Deflate);
+}
+
+bool CompressionCodecDeflate::isAsyncSupported() const
+{
+    return hwCodec->hwEnabled;
+}
+
+void CompressionCodecDeflate::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash);
+}
+
+#define DEFLATE_COMPRESSBOUND(isize) ((isize) + ((isize) >> 12) + ((isize) >> 14) + ((isize) >> 25) + 13) //Aligned with ZLIB
+uint32_t CompressionCodecDeflate::getMaxCompressedDataSize(uint32_t uncompressed_size) const
+{
+    return DEFLATE_COMPRESSBOUND(uncompressed_size);
+}
+
+uint32_t CompressionCodecDeflate::doCompressDataSW(const char * source, uint32_t source_size, char * dest) const
+{
+    return swCodec->doCompressData(source, source_size, dest, getMaxCompressedDataSize(source_size));
+}
+
+uint32_t CompressionCodecDeflate::doCompressData(const char * source, uint32_t source_size, char * dest) const
+{
+    uint32_t res = 0;
+    if (hwCodec->hwEnabled)
+        res = hwCodec->doCompressData(source, source_size, dest, getMaxCompressedDataSize(source_size));
```

Comment:
IIUC, an uninitialized HW codec will return 0 as res here. This is actually some creative use of the API as "0" means the input was successfully compressed into zero bytes. Should we make HWCodec/SWCodec::doCompressData() return a dedicated return type?

Reply:
Done

0 means fail which may confused, for a alternative, I define a dedicated value: RET_ERROR